### PR TITLE
feat(mqtt-hooks): Add MQTT ingress extension to integrate with IoT and automation systems

### PR DESCRIPTION
**Pull request by visoar**
## Summary

Describe the problem and fix in 2-5 bullets:

- Problem: ingress behavior (dispatch, session policy, and template rendering) lived inside HTTP hooks and was partially duplicated, making protocol-level reuse error-prone.
- Why it matters: duplicated ingress behavior risks policy drift and inconsistent security semantics across integrations.
- What changed: extracted shared ingress primitives in gateway core, refactored hooks to consume them, and added an official `@openclaw/mqtt-hooks` extension that maps MQTT events to existing `wake` and `agent` actions.
- What did NOT change (scope boundary): no new action type, no transform runtime, and no changes to unrelated channel delivery pipelines.

## Change Type (select all)

- [ ] Bug fix
- [x] Feature
- [x] Refactor
- [ ] Docs
- [x] Security hardening
- [x] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [x] Integrations
- [x] API / contracts
- [x] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #
- Related #

## User-visible / Behavior Changes

- Adds a new official extension package: `@openclaw/mqtt-hooks`.
- MQTT topics can now map to existing ingress actions (`wake`, `agent`) with bounded payload size, bounded queue/concurrency, startup retained-message guard, and dedupe window.
- HTTP hooks and MQTT ingress now share the same agent/session policy resolution path.

## Security Impact (required)

- New permissions/capabilities? (`Yes`)
- Secrets/tokens handling changed? (`No`)
- New/changed network calls? (`Yes`)
- Command/tool execution surface changed? (`No`)
- Data access scope changed? (`No`)
- If any `Yes`, explain risk + mitigation:
  - The MQTT extension introduces broker connectivity for plugin runtime.
  - Mitigations: payload bound, queue/concurrency bound, retained startup guard, dedupe window, and reuse of existing hook policy gates (`allowedAgentIds`, session key policy).

### Environment

- OS: macOS (local dev)
- Runtime/container: Node 22 + pnpm workspace
- Model/provider: N/A
- Integration/channel (if any): MQTT (plugin tests)
- Relevant config (redacted): plugin test fixtures under `extensions/mqtt-hooks/src/*.test.ts`

### Steps

1. Enable `mqtt-hooks` with broker/runtime/subscription config.
2. Start gateway and publish messages on configured topics.
3. Observe `wake`/`agent` dispatch and policy behavior.

### Expected

- MQTT subscriptions route events through the same ingress policy and dispatch semantics used by hooks.

### Actual

- Behavior matches expectation; validated by targeted tests and full repo checks.

## Evidence

Attach at least one:

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Verification commands run:

- `pnpm test -- extensions/mqtt-hooks/src/config.test.ts extensions/mqtt-hooks/src/dispatch.test.ts extensions/mqtt-hooks/src/envelope.test.ts extensions/mqtt-hooks/src/service.test.ts src/plugin-sdk/subpaths.test.ts`
- `pnpm check`

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios:
  - Shared ingress extraction compiles and passes repository checks.
  - MQTT config parsing/defaults and dispatch behavior.
  - MQTT service lifecycle (connect/subscribe/dedupe/startup retained guard/stop idempotency).
- Edge cases checked:
  - Duplicate subscription ids are rejected.
  - Invalid template field combinations per action are rejected.
  - Agent allowlist rejection path for MQTT agent dispatch.
- What you did **not** verify:
  - Live external broker end-to-end.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes`)
- Config/env changes? (`Yes`)
- Migration needed? (`No`)
- If yes, exact upgrade steps:
  - Optional: install/enable `@openclaw/mqtt-hooks` and add plugin config.

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly:
  - Disable the `mqtt-hooks` plugin entry, or revert this PR.
- Files/config to restore:
  - `plugins.entries["mqtt-hooks"]` and related plugin config.
- Known bad symptoms reviewers should watch for:
  - Queue saturation warnings or subscription schema validation failures.

## Risks and Mitigations

List only real risks for this PR. Add/remove entries as needed. If none, write `None`.

- Risk: MQTT broker instability can increase ingress noise.
  - Mitigation: reconnect path in client + bounded queue/concurrency + dedupe + warning logs.
- Risk: ingress policy drift between protocols.
  - Mitigation: shared gateway ingress primitives are now used by both hooks and MQTT mapping.

### CODE REVIEW

File: extensions/mqtt-hooks/src/service.ts [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+      client = clientFactory(params.pluginConfig.broker);
+
+      client.on("connect", () => {
+        ctx.logger.info(
+          `mqtt-hooks: connected to ${brokerLabel} with ${activeSubscriptions.length} subscription(s)`,
+        );
+        const armStartupRetainedGuard = !hasConnectedOnce;
+        hasConnectedOnce = true;
+        void subscribeAll(armStartupRetainedGuard).catch((err) => {
```
[the remark this answers is not in the document]
Comment (visoar) [reply extensions/mqtt-hooks/src/service.ts]: Great catch. Fixed in 05394d8fd: `hasConnectedOnce` is now set only after the first `subscribeAll(...)` succeeds, so a failed initial subscribe does not permanently keep startup retained guard active. Added regression coverage in `extensions/mqtt-hooks/src/service.test.ts` for the exact failure/reconnect path. Re-ran targeted tests and `pnpm check`.

File: extensions/mqtt-hooks/package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@openclaw/mqtt-hooks",
+  "version": "2026.3.9",
+  "description": "OpenClaw MQTT ingress plugin",
+  "type": "module",
+  "dependencies": {
+    "mqtt": "^5.15.0"
+  },
+  "peerDependencies": {
+    "openclaw": ">=2026.3.8"
```
[the remark this answers is not in the document]
Comment (visoar) [reply extensions/mqtt-hooks/package.json]: Good catch. I raised the peer floor to openclaw >=2026.3.9 because this plugin imports openclaw/plugin-sdk/mqtt-hooks, and I also added devDependencies.openclaw=workspace:* so the workspace still installs cleanly before 2026.3.9 is available on npm. Re-ran pnpm install --lockfile-only and pnpm check locally after this change.

**Comment by visoar** [comment]
Addressed the remaining MQTT review point (dedupe-before-queue) in commit ae09768a0.

What changed:
- `extensions/mqtt-hooks/src/service.ts`
  - compute dedupe key before enqueue
  - skip enqueue when key is already queued (`queuedDedupeKeys`) or already in dedupe cache (`dedupe.peek`)
  - keep dedupe check in task execution path and always release queued keys in `finally`
  - clear queued keys during `stop()` to avoid stale reservations
- `extensions/mqtt-hooks/src/dedupe.ts`
  - expose `peek(...)` for side-effect-free pre-enqueue checks
- `extensions/mqtt-hooks/src/service.test.ts`
  - add regression test: duplicate burst no longer starves a later unique message under constrained concurrency

Validation:
- `pnpm test -- extensions/mqtt-hooks`
- `pnpm check`
